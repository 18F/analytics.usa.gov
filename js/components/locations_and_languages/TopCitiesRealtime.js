import React, { useRef, useEffect } from "react";
import PropTypes from "prop-types";
import d3 from "d3";

import renderBlock from "../../lib/chart_helpers/renderblock";
import transformers from "../../lib/chart_helpers/transformers";

function TopCitiesRealtime({ dataHrefBase }) {
  const dataURL = `${dataHrefBase}/top-cities-realtime.json`;
  const ref = useRef(null);

  useEffect(() => {
    const initRealtimeCitiesChart = async () => {
      const result = await d3
        .select(ref.current)
        .datum({
          source: dataURL,
          block: ref.current,
        })
        .call(
          renderBlock.buildBarChartWithLabel((d) => {
            const cityList = d.data;
            const cityListFiltered = cityList.filter(
              (c) => c.city !== "(not set)" && c.city !== "zz",
            );
            const proportions = transformers.findProportionsOfMetric(
              cityListFiltered,
              (list) => list.map((x) => x.active_visitors),
            );
            return proportions.slice(0, 13);
          }, "city"),
        );
      return result;
    };
    initRealtimeCitiesChart().catch(console.error);
  }, []);

  return (
    <figure
      id="chart_top-cities-90-days"
      data-source={dataURL}
      data-refresh="15"
      ref={ref}
    >
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopCitiesRealtime.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
};

export default TopCitiesRealtime;
