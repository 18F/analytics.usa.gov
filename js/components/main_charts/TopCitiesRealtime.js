import React, { useRef, useEffect, useState } from "react";
import PropTypes from "prop-types";

import ChartBuilder from "../../lib/chart_helpers/chart_builder";
import DataLoader from "../../lib/data_loader";
import transformers from "../../lib/chart_helpers/transformers";

/**
 * Retrieves the realtime top cities report from the passed data URL and
 * creates a visualization for the count of users visiting sites for the current
 * agency from each city.
 *
 * @param {object} props the properties for the component
 * @param {string} props.dataHrefBase the URL of the base location of the data
 * to be downloaded including the agency path. In production this is proxied and
 * redirected to the S3 bucket URL.
 * @param {number} props.refreshSeconds the number of seconds to wait before
 * refreshing chart data.
 * @returns {import('react').ReactElement} The rendered element
 */
function TopCitiesRealtime({ dataHrefBase, refreshSeconds }) {
  const dataURL = `${dataHrefBase}/top-cities-realtime.json`;
  const ref = useRef(null);
  const [realtimeCitiesData, setRealtimeCitiesData] = useState(null);

  useEffect(() => {
    const initRealtimeCitiesChart = async () => {
      if (!realtimeCitiesData) {
        const data = await DataLoader.loadJSON(dataURL);
        await setRealtimeCitiesData(data);
        // Refresh data every interval. useEffect will run and update the chart
        // when the state is changed.
        setInterval(() => {
          DataLoader.loadJSON(dataURL).then((data) => {
            setRealtimeCitiesData(data);
          });
        }, refreshSeconds * 1000);
      } else {
        const chartBuilder = new ChartBuilder();
        await chartBuilder.buildBarChartWithLabel(
          ref.current,
          realtimeCitiesData,
          (d) => {
            const cityList = d.data;
            const cityListFiltered = cityList.filter(
              (c) => c.city !== "(not set)" && c.city !== "zz",
            );
            const proportions = transformers.findProportionsOfMetric(
              cityListFiltered,
              (list) => list.map((x) => x.active_visitors),
            );
            return proportions.slice(0, 13);
          },
          "city",
        );
      }
    };
    initRealtimeCitiesChart().catch(console.error);
  }, [realtimeCitiesData]);

  return (
    <figure id="chart_top-cities-realtime" ref={ref}>
      <div className="data bar-chart"></div>
    </figure>
  );
}

TopCitiesRealtime.propTypes = {
  dataHrefBase: PropTypes.string.isRequired,
  refreshSeconds: PropTypes.number.isRequired,
};

export default TopCitiesRealtime;
