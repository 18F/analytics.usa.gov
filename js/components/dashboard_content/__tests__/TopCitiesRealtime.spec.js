import React from "react";
import { render, screen, waitFor } from "@testing-library/react";

import { delay } from "../../../../spec/support/test_utilities";
import DataLoader from "../../../lib/data_loader";
import TopCitiesRealtime from "../TopCitiesRealtime";

jest.mock("../../../lib/data_loader", () => ({
  ...jest.requireActual("../../../lib/data_loader"),
  loadJSON: jest.fn(),
}));

describe("TopCitiesRealtime", () => {
  let component;
  let data;

  describe("when data is not loaded", () => {
    beforeEach(() => {
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(null);
      });
      component = render(
        <TopCitiesRealtime
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in loading state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data is loaded", () => {
    beforeEach(async () => {
      data = {
        name: "top-cities-realtime",
        query: {},
        meta: {
          name: "Top Cities (Live)",
          description: "Top cities for active onsite users.",
        },
        data: [
          {
            city: "New York",
            active_visitors: "31565",
          },
          {
            city: "Chicago",
            active_visitors: "22808",
          },
          {
            city: "Atlanta",
            active_visitors: "14612",
          },
          {
            city: "Dallas",
            active_visitors: "14457",
          },
          {
            city: "Ashburn",
            active_visitors: "13372",
          },
          {
            city: "Los Angeles",
            active_visitors: "10543",
          },
          {
            city: "Washington",
            active_visitors: "10367",
          },
          {
            city: "Houston",
            active_visitors: "8454",
          },
          {
            city: "Seattle",
            active_visitors: "8418",
          },
          {
            city: "San Jose",
            active_visitors: "5897",
          },
          {
            city: "Boston",
            active_visitors: "5807",
          },
          {
            city: "Miami",
            active_visitors: "5524",
          },
          {
            city: "Reston",
            active_visitors: "5511",
          },
          {
            city: "Philadelphia",
            active_visitors: "5327",
          },
        ],
        totals: {},
        taken_at: "2024-01-05T16:05:48.568Z",
      };

      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.resolve(data);
      });
      component = render(
        <TopCitiesRealtime
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
      await waitFor(() => screen.getByText("Washington"));
      // Wait for barchart transition animation to complete (200 ms, set in
      // js/lib/chart_helpers/barchart.js)
      await delay(300);
    });

    it("renders a component with data loaded", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });
  });

  describe("when data loading has an error", () => {
    const error = "you broke it";

    beforeEach(() => {
      console.error = jest.fn();
      DataLoader.loadJSON.mockImplementation(() => {
        return Promise.reject(error);
      });
      component = render(
        <TopCitiesRealtime
          dataHrefBase="http://www.example.com/data/"
          refreshSeconds={30}
        />,
      );
    });

    it("renders a component in error state", () => {
      expect(component.asFragment()).toMatchSnapshot();
    });

    it("logs the error to console", () => {
      expect(console.error).toHaveBeenCalledWith(error);
    });
  });
});
