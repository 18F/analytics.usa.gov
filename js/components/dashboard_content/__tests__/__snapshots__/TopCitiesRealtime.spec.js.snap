// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TopCitiesRealtime when data is loaded renders a component with data loaded 1`] = `
<DocumentFragment>
  <figure
    id="chart_top-cities-realtime"
  >
    <div
      class="data bar-chart"
    >
      <div
        class="bin"
      >
        <div
          class="label"
        >
          New York
        </div>
        <div
          class="value"
        >
          19.4%
        </div>
        <div
          class="bar"
          style="width: 0.1180199%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Chicago
        </div>
        <div
          class="value"
        >
          14%
        </div>
        <div
          class="bar"
          style="width: 0.08516900000000001%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Atlanta
        </div>
        <div
          class="value"
        >
          9%
        </div>
        <div
          class="bar"
          style="width: 0.0547515%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Dallas
        </div>
        <div
          class="value"
        >
          8.9%
        </div>
        <div
          class="bar"
          style="width: 0.05414315000000001%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Ashburn
        </div>
        <div
          class="value"
        >
          8.2%
        </div>
        <div
          class="bar"
          style="width: 0.0498847%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Los Angeles
        </div>
        <div
          class="value"
        >
          6.5%
        </div>
        <div
          class="bar"
          style="width: 0.03954275%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Washington
        </div>
        <div
          class="value"
        >
          6.4%
        </div>
        <div
          class="bar"
          style="width: 0.03893440000000001%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Houston
        </div>
        <div
          class="value"
        >
          5.2%
        </div>
        <div
          class="bar"
          style="width: 0.0316342%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Seattle
        </div>
        <div
          class="value"
        >
          5.2%
        </div>
        <div
          class="bar"
          style="width: 0.0316342%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          San Jose
        </div>
        <div
          class="value"
        >
          3.6%
        </div>
        <div
          class="bar"
          style="width: 0.021900600000000003%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Boston
        </div>
        <div
          class="value"
        >
          3.6%
        </div>
        <div
          class="bar"
          style="width: 0.021900600000000003%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Miami
        </div>
        <div
          class="value"
        >
          3.4%
        </div>
        <div
          class="bar"
          style="width: 0.0206839%;"
        />
      </div>
      <div
        class="bin"
      >
        <div
          class="label"
        >
          Reston
        </div>
        <div
          class="value"
        >
          3.4%
        </div>
        <div
          class="bar"
          style="width: 0.0206839%;"
        />
      </div>
    </div>
  </figure>
</DocumentFragment>
`;

exports[`TopCitiesRealtime when data is not loaded renders a component in loading state 1`] = `
<DocumentFragment>
  <figure
    id="chart_top-cities-realtime"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;

exports[`TopCitiesRealtime when data loading has an error renders a component in error state 1`] = `
<DocumentFragment>
  <figure
    id="chart_top-cities-realtime"
  >
    <div
      class="data bar-chart"
    />
  </figure>
</DocumentFragment>
`;
